File: orchestrator-go/node.go
// Package orchestrator is the Go parity port of the deterministic node (CAL Exec
// Spec §6/§7 + Canonical Encoding §6.3). It mirrors orchestrator-rs/src/node.rs
// and the TypeScript reference src/node.ts byte-for-byte; see
// ../docs/notes/orchestrator-design.md.
//
// A node folds a program — an ordered list of per-tick blocks, each carrying a
// list of {cal, trace} submissions — through the full pipeline over one evolving
// reducer State:
//
//	for each submission:  cal.created -> cal.signed   (ingress; the reducer enforces
//	                      §6.1 single-in-flight-per-agent and CAL uniqueness)
//	                      validate(cal, snapshot=State, trace) -> lifecycle events
//	                      fold every event through Apply() to advance State
//
// Between blocks it emits tick.advanced. It records the STATE_ROOT after every
// event and the CE §6.3 global stream Merkle root at the end of each tick, and the
// whole event log is byte-for-byte replayable (see Replay). Pure: like the
// validator it consumes execution traces and does not execute steps.
package orchestrator

import (
	"fmt"
	"math/big"

	cal "github.com/paradigm-terra/cal-go"
	calreducer "github.com/paradigm-terra/cal-reducer-go"
	calvalidator "github.com/paradigm-terra/cal-validator-go"
	canonical "github.com/paradigm-terra/canonical-go"
)

// NodeError is a node-level failure (tick regression, a validator-emitted event
// the reducer rejects, or a canonical-encoding error).
type NodeError struct {
	Code   string
	Detail string
}

func (e *NodeError) Error() string { return e.Code + ": " + e.Detail }

func nodeErr(code, detail string) *NodeError { return &NodeError{Code: code, Detail: detail} }
func canonErr(err error) *NodeError {
	return &NodeError{Code: "CANON_ERROR", Detail: fmt.Sprintf("%v", err)}
}

// Submission is one CAL handed to the node with the off-chain executor's trace.
type Submission struct {
	Cal   canonical.Value
	Trace calvalidator.ExecutionTrace
}

// TickBlock is all submissions that land at one tick (must be ≥ the node's tick).
type TickBlock struct {
	Tick        *big.Int
	Submissions []Submission
}

// Program is the start state plus the ordered per-tick blocks to fold.
type Program struct {
	GenesisState canonical.Value
	Ticks        []TickBlock
}

// SubmissionResult records one submission's outcome.
type SubmissionResult struct {
	CalHash string
	AgentID string
	// TerminalStage is nil when rejected at ingress (before validation).
	TerminalStage *string
	// ReasonCode is non-nil only for a FAILED terminal stage.
	ReasonCode *string
	EventTypes []string
	// StateRoots holds the 0x-hex STATE_ROOT after each recorded event.
	StateRoots   []string
	IngressError *string
}

// TickResult is the end-of-tick checkpoint: per-submission results, the STATE_ROOT,
// and the CE §6.3 global stream Merkle root.
type TickResult struct {
	Tick             *big.Int
	Submissions      []SubmissionResult
	StateRoot        string
	GlobalMerkleRoot string
}

// Transcript is the full run output: per-tick checkpoints, the ordered event log
// (fold it from GenesisState to reproduce every root), and the final STATE_ROOT.
type Transcript struct {
	GenesisState   canonical.Value
	Ticks          []TickResult
	EventLog       []canonical.Value
	FinalStateRoot string
}

func hex32(b [32]byte) string { return canonical.ToHexPrefixed(b[:]) }

func eventType(ev canonical.Value) string {
	if o, ok := ev.(*canonical.Object); ok {
		if v, ok := o.Get("event_type"); ok {
			if s, ok := v.(string); ok {
				return s
			}
		}
	}
	return "?"
}

func currentTickOf(state canonical.Value) *big.Int {
	o, ok := state.(*canonical.Object)
	if !ok {
		return big.NewInt(0)
	}
	tv, ok := o.Get("tick")
	if !ok {
		return big.NewInt(0)
	}
	to, ok := tv.(*canonical.Object)
	if !ok {
		return big.NewInt(0)
	}
	if cv, ok := to.Get("current"); ok {
		if iv, ok := cv.(canonical.Int); ok {
			if n, good := new(big.Int).SetString(string(iv), 10); good {
				return n
			}
		}
	}
	return big.NewInt(0)
}

func strField(obj canonical.Value, key string) (string, *NodeError) {
	if o, ok := obj.(*canonical.Object); ok {
		if v, ok := o.Get(key); ok {
			if s, ok := v.(string); ok {
				return s, nil
			}
		}
	}
	return "", nodeErr("BAD_CAL", key+" must be a string")
}

// globalMerkleRoot is the CE §6.3 global Merkle root over a single "global" stream
// (v0.1.0; the constitution's multi-stream list drops in later).
func globalMerkleRoot(state canonical.Value, log []canonical.Value) (string, *NodeError) {
	var lastEventHash [32]byte
	if len(log) > 0 {
		h, err := cal.EventHash(log[len(log)-1])
		if err != nil {
			return "", canonErr(err)
		}
		lastEventHash = h
	}
	stateHash, err := calreducer.StateRootOf(state)
	if err != nil {
		return "", canonErr(err)
	}
	leaf := canonical.StreamLeaf{
		StreamID:      "global",
		StateHash:     stateHash,
		LastEventHash: lastEventHash,
		LastSeqno:     uint64(len(log)),
	}
	root, err := canonical.StreamTreeRoot([]canonical.StreamLeaf{leaf})
	if err != nil {
		return "", canonErr(err)
	}
	return hex32(root), nil
}

// traceAt re-points a trace at the node's tick (a submission must not misreport the
// tick to dodge expiration); everything else is carried through unchanged.
func traceAt(src calvalidator.ExecutionTrace, tick *big.Int) calvalidator.ExecutionTrace {
	return calvalidator.ExecutionTrace{
		CurrentTick:         new(big.Int).Set(tick),
		Steps:               src.Steps,
		StateBefore:         src.StateBefore,
		StateAfter:          src.StateAfter,
		OperatorSigPresent:  src.OperatorSigPresent,
		OwnerSigPresent:     src.OwnerSigPresent,
		PinnedMCPSchemaHash: src.PinnedMCPSchemaHash,
	}
}

// Run folds a program to a transcript. It errors on a tick regression or on a
// validator-emitted event the reducer rejects (an integration defect).
func Run(program *Program) (*Transcript, *NodeError) {
	genesisState := program.GenesisState
	state := genesisState
	var log []canonical.Value
	var ticks []TickResult
	currentTick := currentTickOf(state)

	for _, block := range program.Ticks {
		if block.Tick.Cmp(currentTick) < 0 {
			return nil, nodeErr("TICK_REGRESSION", fmt.Sprintf("block tick %s < current %s", block.Tick, currentTick))
		}
		if block.Tick.Cmp(currentTick) > 0 {
			adv := canonical.NewObject(
				canonical.P("event_type", "tick.advanced"),
				canonical.P("new_tick", canonical.Int(block.Tick.String())),
			)
			ns, aerr := calreducer.Apply(state, adv)
			if aerr != nil {
				return nil, nodeErr("TICK_REJECTED", aerr.Code)
			}
			state = ns
			log = append(log, adv)
			currentTick = new(big.Int).Set(block.Tick)
		}

		var subs []SubmissionResult
		for _, sub := range block.Submissions {
			ch, err := cal.CalHash(sub.Cal)
			if err != nil {
				return nil, canonErr(err)
			}
			calHashHex := canonical.ToHexPrefixed(ch[:])
			agentID, ferr := strField(sub.Cal, "agent_id")
			if ferr != nil {
				return nil, ferr
			}
			var eventTypes []string
			var stateRoots []string

			// Ingress: cal.created then cal.signed (reducer enforces §6.1 / uniqueness).
			var ingressError *string
			ingress := []canonical.Value{
				canonical.NewObject(canonical.P("event_type", "cal.created"), canonical.P("cal_hash", calHashHex), canonical.P("agent_id", agentID)),
				canonical.NewObject(canonical.P("event_type", "cal.signed"), canonical.P("cal_hash", calHashHex)),
			}
			for _, ev := range ingress {
				ns, aerr := calreducer.Apply(state, ev)
				if aerr != nil {
					code := aerr.Code
					ingressError = &code
					break
				}
				state = ns
				log = append(log, ev)
				eventTypes = append(eventTypes, eventType(ev))
				sr, serr := calreducer.StateRootOf(state)
				if serr != nil {
					return nil, canonErr(serr)
				}
				stateRoots = append(stateRoots, hex32(sr))
			}
			if ingressError != nil {
				subs = append(subs, SubmissionResult{CalHash: calHashHex, AgentID: agentID, EventTypes: eventTypes, StateRoots: stateRoots, IngressError: ingressError})
				continue
			}

			// Validate against the live state (tick pinned), then fold the events.
			trace := traceAt(sub.Trace, currentTick)
			res, verr := calvalidator.Validate(sub.Cal, calHashHex, state, trace)
			if verr != nil {
				return nil, nodeErr("VALIDATE_ERROR", verr.Error())
			}
			for _, ev := range res.Events {
				ns, aerr := calreducer.Apply(state, ev)
				if aerr != nil {
					return nil, nodeErr("APPLY_FAILED", fmt.Sprintf("%s event %s rejected: %s", res.TerminalStage, eventType(ev), aerr.Code))
				}
				state = ns
				log = append(log, ev)
				eventTypes = append(eventTypes, eventType(ev))
				sr, serr := calreducer.StateRootOf(state)
				if serr != nil {
					return nil, canonErr(serr)
				}
				stateRoots = append(stateRoots, hex32(sr))
			}

			stage := res.TerminalStage
			var reason *string
			if res.ReasonCode != "" {
				rc := res.ReasonCode
				reason = &rc
			}
			subs = append(subs, SubmissionResult{
				CalHash:       calHashHex,
				AgentID:       agentID,
				TerminalStage: &stage,
				ReasonCode:    reason,
				EventTypes:    eventTypes,
				StateRoots:    stateRoots,
			})
		}

		stateRoot, serr := calreducer.StateRootOf(state)
		if serr != nil {
			return nil, canonErr(serr)
		}
		gmr, gerr := globalMerkleRoot(state, log)
		if gerr != nil {
			return nil, gerr
		}
		ticks = append(ticks, TickResult{
			Tick:             new(big.Int).Set(currentTick),
			Submissions:      subs,
			StateRoot:        hex32(stateRoot),
			GlobalMerkleRoot: gmr,
		})
	}

	fsr, ferr := calreducer.StateRootOf(state)
	if ferr != nil {
		return nil, canonErr(ferr)
	}
	return &Transcript{GenesisState: genesisState, Ticks: ticks, EventLog: log, FinalStateRoot: hex32(fsr)}, nil
}

// Replay re-folds an event log from a start state and returns the final STATE_ROOT (§7.2).
func Replay(eventLog []canonical.Value, genesisState canonical.Value) (string, *NodeError) {
	state := genesisState
	for _, ev := range eventLog {
		ns, aerr := calreducer.Apply(state, ev)
		if aerr != nil {
			return "", nodeErr("REPLAY_FAILED", aerr.Code)
		}
		state = ns
	}
	sr, err := calreducer.StateRootOf(state)
	if err != nil {
		return "", canonErr(err)
	}
	return hex32(sr), nil
}

File: orchestrator-rs/src/node.rs
//! The deterministic node (CAL Exec Spec §6/§7 + Canonical Encoding §6.3). Rust
//! parity port of `src/node.ts`; see the crate docs and `docs/notes/orchestrator-design.md`.

use paradigm_terra_cal::{cal_hash, event_hash};
use paradigm_terra_cal_gas::U256;
use paradigm_terra_cal_reducer::{apply, state_root_of};
use paradigm_terra_cal_validator::{validate, ExecutionTrace, StepResult};
use paradigm_terra_canonical::integers::to_hex_prefixed;
use paradigm_terra_canonical::jcs::JcsValue;
use paradigm_terra_canonical::merkle::{stream_tree_root, StreamLeaf};
use paradigm_terra_canonical::CanonicalError;

#[derive(Debug, Clone)]
pub struct NodeError {
    pub code: &'static str,
    pub detail: String,
}

fn node_err(code: &'static str, detail: impl Into<String>) -> NodeError {
    NodeError { code, detail: detail.into() }
}
fn canon_err(e: CanonicalError) -> NodeError {
    NodeError { code: "CANON_ERROR", detail: format!("{e:?}") }
}

pub struct Submission {
    pub cal: JcsValue,
    pub trace: ExecutionTrace,
}
pub struct TickBlock {
    pub tick: U256,
    pub submissions: Vec<Submission>,
}
pub struct Program {
    pub genesis_state: JcsValue,
    pub ticks: Vec<TickBlock>,
}

#[derive(Debug, Clone)]
pub struct SubmissionResult {
    pub cal_hash: String,
    pub agent_id: String,
    /// `None` when rejected at ingress (cal.created/cal.signed) before validation.
    pub terminal_stage: Option<String>,
    pub reason_code: Option<String>,
    pub event_types: Vec<String>,
    /// STATE_ROOT (0x-hex) after each recorded event.
    pub state_roots: Vec<String>,
    pub ingress_error: Option<String>,
}
pub struct TickResult {
    pub tick: U256,
    pub submissions: Vec<SubmissionResult>,
    pub state_root: String,
    pub global_merkle_root: String,
}
pub struct Transcript {
    pub genesis_state: JcsValue,
    pub ticks: Vec<TickResult>,
    pub event_log: Vec<JcsValue>,
    pub final_state_root: String,
}

fn hex32(b: &[u8; 32]) -> String {
    to_hex_prefixed(b)
}

fn current_tick_of(state: &JcsValue) -> U256 {
    match state.get("tick").and_then(|t| t.get("current")) {
        Some(JcsValue::Int(s)) => U256::from_dec_str(s).unwrap_or(U256::ZERO),
        _ => U256::ZERO,
    }
}

fn str_field(obj: &JcsValue, key: &str) -> Result<String, NodeError> {
    match obj.get(key) {
        Some(JcsValue::Str(s)) => Ok(s.clone()),
        _ => Err(node_err("BAD_CAL", format!("{key} must be a string"))),
    }
}

/// CE §6.3 global Merkle root over a single `"global"` stream (v0.1.0; the
/// constitution's multi-stream list drops in later).
fn global_merkle_root(state: &JcsValue, log: &[JcsValue]) -> Result<String, NodeError> {
    let last_event_hash = match log.last() {
        Some(ev) => event_hash(ev).map_err(canon_err)?,
        None => [0u8; 32],
    };
    let leaf = StreamLeaf {
        stream_id: "global".to_string(),
        state_hash: state_root_of(state).map_err(canon_err)?,
        last_event_hash,
        last_seqno: log.len() as u64,
    };
    Ok(hex32(&stream_tree_root(&[leaf]).map_err(canon_err)?))
}

fn obj(pairs: Vec<(&str, JcsValue)>) -> JcsValue {
    JcsValue::object(pairs)
}

/// Re-run `validate` against the live state with `current_tick` pinned to the node's
/// tick (a submission must not misreport the tick to dodge expiration).
fn trace_at(src: &ExecutionTrace, tick: &U256) -> ExecutionTrace {
    ExecutionTrace {
        current_tick: tick.clone(),
        steps: src
            .steps
            .iter()
            .map(|s| StepResult { ok: s.ok, effects: s.effects.clone(), error_detail: s.error_detail.clone() })
            .collect(),
        state_before: src.state_before.clone(),
        state_after: src.state_after.clone(),
        owner_sig_present: src.owner_sig_present,
        pinned_mcp_schema_hash: src.pinned_mcp_schema_hash.clone(),
    }
}

/// Run a program to a transcript. Errors on a tick regression or a validator-emitted
/// event the reducer rejects (an integration defect).
pub fn run(program: &Program) -> Result<Transcript, NodeError> {
    let genesis_state = program.genesis_state.clone();
    let mut state = genesis_state.clone();
    let mut log: Vec<JcsValue> = Vec::new();
    let mut ticks: Vec<TickResult> = Vec::new();
    let mut current_tick = current_tick_of(&state);

    for block in &program.ticks {
        if block.tick < current_tick {
            return Err(node_err("TICK_REGRESSION", format!("block tick {} < current {}", block.tick.to_dec_str(), current_tick.to_dec_str())));
        }
        if block.tick > current_tick {
            let adv = obj(vec![("event_type", JcsValue::string("tick.advanced")), ("new_tick", JcsValue::Int(block.tick.to_dec_str()))]);
            state = apply(&state, &adv).map_err(|e| node_err("TICK_REJECTED", e.code))?;
            log.push(adv);
            current_tick = block.tick.clone();
        }

        let mut subs: Vec<SubmissionResult> = Vec::new();
        for sub in &block.submissions {
            let cal_hash_hex = to_hex_prefixed(&cal_hash(&sub.cal).map_err(canon_err)?);
            let agent_id = str_field(&sub.cal, "agent_id")?;
            let mut event_types: Vec<String> = Vec::new();
            let mut state_roots: Vec<String> = Vec::new();

            // Ingress: cal.created then cal.signed (reducer enforces §6.1 / uniqueness).
            let mut ingress_error: Option<String> = None;
            for ev in [
                obj(vec![("event_type", JcsValue::string("cal.created")), ("cal_hash", JcsValue::string(&cal_hash_hex)), ("agent_id", JcsValue::string(&agent_id))]),
                obj(vec![("event_type", JcsValue::string("cal.signed")), ("cal_hash", JcsValue::string(&cal_hash_hex))]),
            ] {
                match apply(&state, &ev) {
                    Ok(ns) => {
                        state = ns;
                        let et = ev.get("event_type").and_then(JcsValue::as_str).unwrap_or("?").to_string();
                        log.push(ev);
                        event_types.push(et);
                        state_roots.push(hex32(&state_root_of(&state).map_err(canon_err)?));
                    }
                    Err(e) => {
                        ingress_error = Some(e.code.to_string());
                        break;
                    }
                }
            }
            if let Some(code) = ingress_error {
                subs.push(SubmissionResult { cal_hash: cal_hash_hex, agent_id, terminal_stage: None, reason_code: None, event_types, state_roots, ingress_error: Some(code) });
                continue;
            }

            // Validate against the live state (tick pinned), then fold the events.
            let trace = trace_at(&sub.trace, &current_tick);
            let res = validate(&sub.cal, &cal_hash_hex, &state, &trace).map_err(|e| node_err("VALIDATE_ERROR", format!("{e:?}")))?;
            for ev in &res.events {
                match apply(&state, ev) {
                    Ok(ns) => state = ns,
                    Err(e) => return Err(node_err("APPLY_FAILED", format!("{} event {} rejected: {}", res.terminal_stage, ev.get("event_type").and_then(JcsValue::as_str).unwrap_or("?"), e.code))),
                }
                log.push(ev.clone());
                event_types.push(ev.get("event_type").and_then(JcsValue::as_str).unwrap_or("?").to_string());
                state_roots.push(hex32(&state_root_of(&state).map_err(canon_err)?));
            }
            subs.push(SubmissionResult {
                cal_hash: cal_hash_hex,
                agent_id,
                terminal_stage: Some(res.terminal_stage.to_string()),
                reason_code: res.reason_code.map(str::to_string),
                event_types,
                state_roots,
                ingress_error: None,
            });
        }

        ticks.push(TickResult {
            tick: current_tick.clone(),
            submissions: subs,
            state_root: hex32(&state_root_of(&state).map_err(canon_err)?),
            global_merkle_root: global_merkle_root(&state, &log)?,
        });
    }

    let final_state_root = hex32(&state_root_of(&state).map_err(canon_err)?);
    Ok(Transcript { genesis_state, ticks, event_log: log, final_state_root })
}

/// Re-fold an event log from a start state and return the final STATE_ROOT (§7.2).
pub fn replay(event_log: &[JcsValue], genesis_state: &JcsValue) -> Result<String, NodeError> {
    let mut state = genesis_state.clone();
    for ev in event_log {
        state = apply(&state, ev).map_err(|e| node_err("REPLAY_FAILED", e.code))?;
    }
    Ok(hex32(&state_root_of(&state).map_err(canon_err)?))
}
